crates/p2p/src/tests/common.rs

```diff
@@ -1,57 +1,108 @@
 //! Helper functions for the P2P tests.
 
-use std::time::Duration;
+use std::{sync::Once, time::Duration};
 
 use futures::future::join_all;
-use libp2p::{Multiaddr, PeerId, build_multiaddr, identity::Keypair};
+use libp2p::{
+    Multiaddr, PeerId, build_multiaddr,
+    identity::{Keypair, PublicKey},
+};
 use rand::Rng;
 use tokio_util::{sync::CancellationToken, task::TaskTracker};
 use tracing::debug;
+use tracing_subscriber::{EnvFilter, fmt, layer::SubscriberExt, util::SubscriberInitExt};
 
 #[cfg(feature = "request-response")]
 use crate::swarm::handle::ReqRespHandle;
-use crate::swarm::{
-    self, P2P, P2PConfig,
-    handle::{CommandHandle, GossipHandle},
+use crate::{
+    signer::ApplicationSigner,
+    swarm::{
+        self, P2P, P2PConfig,
+        handle::{CommandHandle, GossipHandle},
+    },
 };
 
-pub(crate) struct User {
-    pub(crate) p2p: P2P,
+/// Only attempt to start tracing once
+///
+/// it is needed for supporting plain `cargo test`
+pub(crate) fn init_tracing() {
+    static INIT: Once = Once::new();
+
+    INIT.call_once(|| {
+        tracing_subscriber::registry()
+            .with(EnvFilter::from_default_env())
+            .with(fmt::layer().with_file(true).with_line_number(true))
+            .init();
+    });
+}
+
+/// Mock ApplicationSigner for testing that stores the actual keypair
+#[derive(Debug, Clone)]
+pub(crate) struct MockApplicationSigner {
+    // Store the actual keypair that corresponds to the app public key
+    app_keypair: Keypair,
+}
+
+impl MockApplicationSigner {
+    pub(crate) const fn new(app_keypair: Keypair) -> Self {
+        Self { app_keypair }
+    }
+}
+
+impl ApplicationSigner for MockApplicationSigner {
+    fn sign(
+        &self,
+        message: &[u8],
+        _app_public_key: PublicKey,
+    ) -> Result<Vec<u8>, Box<dyn std::error::Error + Send + Sync>> {
```

Comment:
Can we make this not a `Vec<u8>` and actually a `[u8; 64]` like we did with the `signature` field in the `SignedMessage` struct?